Move node_tilable_track_nums_ to rr_graph_storage and add remove_node

amin1377 opened the pull request:
**This PR depends on PR #3342. Please wait until that PR is merged.**

In this PR, node_tilable_track_nums_ is moved into the RR graph storage instead of residing in the RR graph builder, since this data structure (when using a tileable RR graph) needs to remain valid for the entire lifetime of the RR graph.

Additionally, a method is added to facilitate removing nodes from the RR graph.

AmirhosseinPoolad commented on libs/librrgraph/src/base/rr_graph_storage.cpp:
```diff
@@ -625,6 +629,54 @@ void t_rr_graph_storage::set_node_direction(RRNodeId id, Direction new_direction
+void t_rr_graph_storage::set_node_ptc_nums(RRNodeId node, const std::string& ptc_str) {
+    VTR_ASSERT(size_t(node) < node_storage_.size());
+    std::vector<std::string> ptc_tokens = vtr::StringToken(ptc_str).split(",");
+    VTR_ASSERT(ptc_tokens.size() >= 1);
```
VTR_ASSERT(!ptc_tokens.empty());

AmirhosseinPoolad commented on libs/librrgraph/src/base/rr_graph_storage.cpp:
```diff
@@ -651,6 +703,80 @@ void t_rr_graph_storage::set_virtual_clock_network_root_idx(RRNodeId virtual_clo
+    }
+
+    std::vector<RREdgeId> removed_edges;
+    auto adjust_edges = [&](vtr::vector<RREdgeId, RRNodeId>& edge_nodes) {
```
Please explicitly add the variables you want to the capture list.

AmirhosseinPoolad commented on libs/librrgraph/src/base/rr_graph_storage.cpp:
```diff
@@ -651,6 +703,80 @@ void t_rr_graph_storage::set_virtual_clock_network_root_idx(RRNodeId virtual_clo
+    // If a node ID is larger than the last element, its ID should be reduced by 
+    // the total number of nodes being removed.
+    std::vector<RRNodeId> sorted_nodes = nodes;
+    std::sort(sorted_nodes.begin(), sorted_nodes.end());
```
How much do you care about not mutating the list of nodes for removal? This copy seems unnecessary to me but it's your call.

AmirhosseinPoolad commented on libs/librrgraph/src/base/rr_graph_storage.cpp:
```diff
@@ -651,6 +703,80 @@ void t_rr_graph_storage::set_virtual_clock_network_root_idx(RRNodeId virtual_clo
     }
 }
 
+void t_rr_graph_storage::remove_nodes(const std::vector<RRNodeId>& nodes) {
```
Rename "nodes" to something like "rr_nodes_to_remove"

AmirhosseinPoolad commented on libs/librrgraph/src/base/rr_graph_storage.cpp:
```diff
@@ -651,6 +703,80 @@ void t_rr_graph_storage::set_virtual_clock_network_root_idx(RRNodeId virtual_clo
+    };
+
+    adjust_edges(edge_src_node_);
+    adjust_edges(edge_dest_node_);
```
What about edge_switch_ and edge_remapped_?

amin1377 replied:
I need to modify data structures that use RRNodeId (either as a key or value). The two data structures you mentioned only deal with edge IDs.

amin1377 commented:
@AmirhosseinPoolad: I’ve addressed your comments. I’d appreciate it if you could take another look at the PR and let me know if you have any further feedback. It would be great if we could finalize this PR today!

amin1377 commented on libs/librrgraph/src/base/rr_graph_storage.cpp:
```diff
@@ -178,7 +182,7 @@ bool t_rr_graph_storage::verify_first_edges() const {
 
 void t_rr_graph_storage::init_fan_in() {
     //Reset all fan-ins to zero
-    edges_read_ = true;
```
@soheilshahrouz: If you get a chance, I’d appreciate your thoughts on this. The reason I believe this needs to be removed is that edge_read should only be set when we are modifying data structures related to RR edges; for example, when updating edge_sink_node_ or the data structure that stores the first edge for each node.

In this case, we are only modifying the data structure that tracks the number of fan-ins for each node, and I should be able to call this without requiring edge_read to be set to true (since that would imply that edges must have been partitioned, etc.).

AmirhosseinPoolad reviewed (commented):
Sorry for being late! This looks good other than the lambda capture list (super dangerous imo to capture  *everything* by non-const reference and it wouldn't be clear to future maintainers which outside variables the lambda is actually using without going through the lambda code). I still need another person to look at the flag change.

amin1377 commented:
Thanks, @AmirhosseinPoolad . I’ve applied the changes you recommended. Please let me know if you have any further comments.
